workflow/executor/pns/pns.go

```diff
@@ -0,0 +1,323 @@
+	deadline := time.Now().Add(timeout)
+	for {
+		p.updateCtrIDMap()
+		if p.mainFS != nil {
+			log.Info("Stopped root processes polling due to successful securing of main root fs")
+			break
+		}
+		if time.Now().After(deadline) {
+			log.Warnf("Polling root processes timed out (%v)", timeout)
+			break
+		}
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+func (p *PNSExecutor) GetOutputStream(containerID string, combinedOutput bool) (io.ReadCloser, error) {
+	if !combinedOutput {
+		log.Warn("non combined output unsupported")
+	}
+	opts := v1.PodLogOptions{
+		Container: common.MainContainerName,
+	}
+	return p.clientset.CoreV1().Pods(p.namespace).GetLogs(p.podName, &opts).Stream()
+}
+
+// Kill a list of containerIDs first with a SIGTERM then with a SIGKILL after a grace period
+func (p *PNSExecutor) Kill(containerIDs []string) error {
+	var asyncErr error
+	wg := sync.WaitGroup{}
+	for _, cid := range containerIDs {
+		wg.Add(1)
+		go func(containerID string) {
+			err := p.killContainer(containerID)
+			if err != nil && asyncErr != nil {
+				asyncErr = err
+			}
+			wg.Done()
+		}(cid)
+	}
+	wg.Wait()
+	return asyncErr
+}
+
+func (p *PNSExecutor) killContainer(containerID string) error {
+	pid, err := p.getContainerPID(containerID)
+	if err != nil {
+		return err
+	}
+	// On Unix systems, FindProcess always succeeds and returns a Process
+	// for the given pid, regardless of whether the process exists.
+	proc, _ := os.FindProcess(pid)
+	err = proc.Signal(syscall.SIGTERM)
+
+	waitPIDOpts := executil.WaitPIDOpts{Timeout: time.Second}
+	err = executil.WaitPID(pid, waitPIDOpts)
+	if err == nil {
+		log.Infof("PID %d completed", pid)
+		return nil
+	}
+	if err != executil.ErrWaitPIDTimeout {
+		return err
+	}
+	log.Warnf("Timed out (%v) waiting for pid %d to complete after SIGTERM. Issing SIGKILL", waitPIDOpts.Timeout, pid)
+	return proc.Signal(syscall.SIGKILL)
+}
+
+func (p *PNSExecutor) getContainerPID(containerID string) (int, error) {
+	pid, ok := p.ctrIDToPid[containerID]
+	if ok {
+		return pid, nil
+	}
+	p.updateCtrIDMap()
+	pid, ok = p.ctrIDToPid[containerID]
+	if !ok {
+		return -1, errors.InternalErrorf("Could not find associated pid for containerID %s", containerID)
+	}
+	return pid, nil
+}
+
+// updateCtrIDMap updates the mapping between container IDs to PIDs
+// If openRootFs is true, will open /proc/PID/root for all PIDs which are root PIDs
+func (p *PNSExecutor) updateCtrIDMap() error {
+	allProcs, err := gops.Processes()
+	if err != nil {
+		return errors.InternalWrapError(err)
+	}
+	thisPid := os.Getpid()
+	for _, proc := range allProcs {
+		pid := proc.Pid()
+		if pid == 1 || pid == thisPid || proc.PPid() != 0 {
+			// ignore the pause container, our own pid, and non-root processes
+			continue
+		}
+		if p.mainFS == nil {
+			// if we have yet to secure the main container's filesystem, grab all root filesystems
+			if _, ok := p.pidFileHandles[pid]; !ok {
+				rootPath := fmt.Sprintf("/proc/%d/root", pid)
```

Comment:
This is not possible in my environment.

Could you try the following manifest and try to browse the file system of the main container from the sidecar?


```
kind: Pod
apiVersion: v1
metadata:
  name: pns
spec:
  containers:
  - name: main
    image: ubuntu:16.04
    command: [sleep, infinity]
    securityContext:
      runAsUser: 1234
  - name: sidecar
    image: ubuntu:16.04
    command: [sleep, infinity]
    securityContext:
      runAsUser: 0
  shareProcessNamespace: true
```

Even if you are root, you can't browse it. Do I have to set up something to allow it?

Reply:
I think I'm seeing same behavior. [artifact-passing.yaml](https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml) fails even if I add 'sleep 10' to generate-artifact step. According to logs executor was able to save artifacts and then failed because could not chroot into main:

```
time="2019-02-13T18:46:08Z" level=info msg="Saving output artifacts"
time="2019-02-13T18:46:08Z" level=info msg="Staging artifact: hello-art"
time="2019-02-13T18:46:08Z" level=info msg="Copying /tmp/hello_world.txt from container base image layer to /argo/outputs/artifacts/hello-art.tgz"
time="2019-02-13T18:46:08Z" level=error msg="executor error: could not chroot into main for artifact collection: container may have exited too quickly
```

Reply:
Which version kubernetes? Process namespace sharing is beta on k8s 1.12. On lower k8s versions, it requires a feature gate.

Reply:
I use Kubernetes v1.12.3, but it should be `Process Namespace Sharing is a beta feature that is enabled by default.` based on [the document](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/).

I can see other cotainer's processes by `ps aux` but cannot access the filesystem unless the process is owned by the user.

Reply:
Good catch! If the main container is not the same user, then you are correct that this won't work. The fix is that we need to grant the wait sidecar SYS_PTRACE capabilities.
```
    securityContext:
      capabilities:
        add:
        - SYS_PTRACE
```

Reply:
OK I fixed the SYS_PTRACE issue, and also a timing related issue. PNS should be working much more reliably now.

Reply:
Cool. Let me try it. 👍